include/revng/Pipeline/LLVMContainer.h

```diff
@@ -0,0 +1,201 @@
+    if (ToSet == nullptr)
+      continue;
+    To.setMetadata(index, ToSet);
+  }
+}
+
+/// Implementation that can be derived by anyone so that multiple identical
+/// llvmcontainers can exists so that inspector kinds do not pollute each other.
+template<typename MostDerived>
+class LLVMContainerBase : public InspectableContainer<MostDerived> {
+private:
+  LLVMContainerImpl Container;
+
+public:
+  ~LLVMContainerBase() override {}
+
+  LLVMContainerBase(Context &Ctx, std::unique_ptr<llvm::Module> M) :
+    InspectableContainer<MostDerived>(Ctx), Container(std::move(M)) {}
+
+  const llvm::Module &getModule() const { return Container.getModule(); }
+  llvm::Module &getModule() { return Container.getModule(); }
+
+  void mergeBackImpl(MostDerived &&Other) final {
+    Container.mergeBackDerived(Other.Container);
+  }
+
+  llvm::Error storeToDisk(llvm::StringRef Path) const final {
+    return Container.storeToDisk(Path);
+  }
+
+  llvm::Error loadFromDisk(llvm::StringRef Path) final {
+    return Container.loadFromDisk(Path);
+  }
+
+  std::unique_ptr<ContainerBase>
+  cloneFiltered(const TargetsList &Targets) const final {
+    using InspectorT = LLVMContainerInspector<MostDerived>;
+    auto ToClone = InspectorT::enumerateAllFunctions(Targets, self());
+    auto ToClonedNotOwned = InspectorT::enumerateNonOnwedFunctions(self());
+
+    const auto Filter = [&ToClone, &ToClonedNotOwned](const auto &GlobalSym) {
+      if (not llvm::isa<llvm::Function>(GlobalSym))
+        return true;
+
+      const auto &F = llvm::cast<llvm::Function>(GlobalSym);
+      return ToClone.count(F) != 0 or ToClonedNotOwned.count(F) != 0;
+    };
+
+    llvm::ValueToValueMapTy Map;
+    auto Module = llvm::CloneModule(Container.getModule(), Map, Filter);
+
+    for (const llvm::Function &Original : Container.getModule().functions()) {
+      if (Original.isDeclaration())
+        cloneMetadata<llvm::Function>(Map, &Original);
+    }
+
+    return std::make_unique<MostDerived>(*this->Ctx, std::move(Module));
+  }
+
+  template<typename... LLVMPipePassess>
+  static PipeWrapper
+  wrapLLVMPassess(std::string LLVMModuleName, LLVMPipePassess &&...P) {
+    return PipeWrapper(LLVMPipeBase<MostDerived>(std::move(P)...),
+                       { std::move(LLVMModuleName) });
+  }
+
+  const LLVMContainerImpl &getContainer() const { return Container; }
+
+  LLVMContainerImpl &getContainer() { return Container; }
+
+  MostDerived &self() { return *static_cast<MostDerived *>(this); }
+
+  const MostDerived &self() const {
+    return *static_cast<const MostDerived *>(this);
+  }
+};
+
+/// regular llvm container used in most of the place
+class LLVMContainer : public LLVMContainerBase<LLVMContainer> {
+public:
+  static char ID;
+
+  ~LLVMContainer() override {}
+
+  LLVMContainer(Context &Ctx, std::unique_ptr<llvm::Module> M) :
+    LLVMContainerBase<LLVMContainer>(Ctx, std::move(M)) {}
+};
+
+using LLVMPipe = LLVMPipeBase<LLVMContainer>;
```

Comment:
Wouldn't make more sense to have `LLVMPipe.h` depend on `LLVMContainer.h` than viceversa?
A pipe needs to know about its containers, not viceversa.

Reply:
the issue with it is just due to practicality, this dependecy order allows us to have the method wrapLLVMPassess into the container type so that one can do

```cpp
ContainerT::wrapLLVMPassess(pass1, ..., passN);
```
and this return a pipe with a appropriate type for the current container

if we invert the order wrapLLVMPassess must become a free function and it's a bit less comfortable to use.

Reply:
Uhm, `wrapLLVMPassess` takes template arguments anyway, so there's no real dependency, isn't it?
In any case, let's strive for the most straightforward organization of dependencies between components.
The complexity of the thing is already quite significant.